tap_trello/streams.py

```diff
@@ -76,59 +76,46 @@ def _get_window_state(self):
-        singer.write_bookmark(self.state, self.stream_id, "in_middle_of_sync", True)
+    def on_window_started(self):
+        if singer.get_bookmark(self.state, self.stream_id, 'sub_window_end') is None:
+            if singer.get_bookmark(self.state, self.stream_id, 'window_start') is None:
+                singer.write_bookmark(self.state, self.stream_id, "window_start", self.config['start_date'])
+            now = utils.strftime(utils.now())
+            singer.write_bookmark(self.state, self.stream_id, "window_end", min(self.config.get('end_date', now), now))
         singer.write_state(self.state)
 
-    def on_sync_finished(self):
+    def on_window_finished(self):
         # Upgrade sync_start to a bookmark here to be used next time through to override `start_date`.
-        last_sync_end = singer.get_bookmark(self.state, self.stream_id, "sync_end")
-        singer.write_bookmark(self.state, self.stream_id, "last_sync_end", last_sync_end)
-        singer.clear_bookmark(self.state, self.stream_id, "sync_end")
-        singer.write_bookmark(self.state, self.stream_id, "in_middle_of_sync", False)
+        window_start = singer.get_bookmark(self.state, self.stream_id, "window_end")
+        singer.write_bookmark(self.state, self.stream_id, "window_start", window_start)
+        singer.clear_bookmark(self.state, self.stream_id, "window_end")
         singer.write_state(self.state)
 
     def get_records(self, format_values, params=None):
         """ Overrides the default get_records to provide date_window pagination and bookmarking. """
         if params is None:
             params = {}
 
-        window_start, window_sub_end, next_window_start = self._get_window_state()
+        window_start, sub_window_end, window_end = self._get_window_state()
         window_start -= timedelta(milliseconds=1) # To make start inclusive
 
-        if window_sub_end is not None:
-            # Indicates we need to resume, next_window_start should be present
-            next_window_start -= timedelta(milliseconds=1) # To make next start inclusive
-            LOGGER.info('Paginating with window_sub_end')
-            for rec in self._paginate_window(window_start, window_sub_end, format_values, params):
+        if sub_window_end is not None:
+            for rec in self._paginate_window(window_start, sub_window_end, format_values, params):
                 yield rec
-
         else:
-            # sync_end refers to the right-bound of our last window
-            sync_end = singer.get_bookmark(self.state, self.stream_id, "sync_end")
-            sync_end = utils.strptime_to_utc(sync_end)
-
-            self._update_bookmark("next_window_start", sync_end)
-            for rec in self._paginate_window(window_start, sync_end, format_values, params):
+            for rec in self._paginate_window(window_start, window_end, format_values, params):
                 yield rec
```

Comment:
Maybe this could be `self._paginate_window(window_start, sub_window_end or window_end, format_values, params)`. Depends on how that looks

Reply:
I'm neither here nor there on this, having them split might be more readable.